src/thor/timedistancematrix.cc

```diff
@@ -556,17 +553,33 @@ bool TimeDistanceMatrix::UpdateDestinations(
 }
 
 // Form the time, distance matrix from the destinations list
-std::vector<TimeDistance> TimeDistanceMatrix::FormTimeDistanceMatrix(GraphReader& reader,
-                                                                     const std::string& origin_dt,
-                                                                     const uint64_t& origin_tz,
-                                                                     const GraphId& pred_id) {
-  std::vector<TimeDistance> td;
-  for (auto& dest : destinations_) {
-    auto date_time = get_date_time(origin_dt, origin_tz, pred_id, reader,
-                                   static_cast<uint64_t>(dest.best_cost.secs + .5f));
-    td.emplace_back(dest.best_cost.secs, dest.distance, date_time);
+void TimeDistanceMatrix::FormTimeDistanceMatrix(Api& request,
+                                                GraphReader& reader,
+                                                const bool forward,
+                                                const uint32_t origin_index,
+                                                const std::string& origin_dt,
+                                                const uint64_t& origin_tz,
+                                                const GraphId& pred_id,
+                                                std::vector<std::string>& out_date_times) {
+  // when it's forward, origin_index will be the source_index
+  // when it's reverse, origin_index will be the target_index
+  valhalla::Matrix& matrix = *request.mutable_matrix();
+  for (uint32_t i = 0; i < destinations_.size(); i++) {
+    auto& dest = destinations_[i];
+    float time = dest.best_cost.secs + .5f;
+    auto pbf_idx = forward ? (origin_index * request.options().targets().size()) + i
+                           : (i * request.options().targets().size()) + origin_index;
+    matrix.mutable_from_indices()->Set(pbf_idx, forward ? origin_index : i);
+    matrix.mutable_to_indices()->Set(pbf_idx, forward ? i : origin_index);
+    matrix.mutable_distances()->Set(pbf_idx, dest.distance);
+    matrix.mutable_times()->Set(pbf_idx, time);
+
+    // this logic doesn't work with string repeated fields, gotta collect them
+    // and process them later
+    auto date_time =
+        get_date_time(origin_dt, origin_tz, pred_id, reader, static_cast<uint64_t>(time));
+    out_date_times[pbf_idx] = date_time;
```

Comment:
this pisses me off a bit.. we can't use the `Resize()` and `Set()` syntax for the string repeated fields, protobuf has no such concept for those.. I didn't really know how to do  it any better, so:
- before we expand the sources or targets, we initialize an empty `std::vector<std::string>`, which we pass into this function for every source or target forming
- here we set the date_time string on the right index
- after we're all done with all expansions, we go over this vector and populate the PBF's `date_times`

Any suggestion on how to make it less ugly are welcome.